DOCSP-38861 - OIDC (#81)
mongodb/docs-pymongo: source/security/authentication.txt

The following example shows how to define a callback class named ``MyCallback``. This class
includes a ``fetch()`` method that retrieves an OIDC token from the file specified in the
``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable.

.. literalinclude:: /includes/authentication/aws-eks-mongoclient.py
   :language: python
   :copyable: true
   :start-after: # define callback, properties, and MongoClient
   :end-before: properties = {"OIDC_CALLBACK": MyCallback()}

After you define your callback class, create a Python dictionary that contains one key,
``OIDC_CALLBACK``, whose value is an instance of your custom callback class:

.. literalinclude:: /includes/authentication/aws-eks-mongoclient.py
   :language: python
   :copyable: true
   :start-after: return OIDCCallbackResult(access_token=token)
   :end-before: client = MongoClient(

Finally, set the following connection options:

- ``authMechanism``: Set to ``"MONGODB-OIDC"``.
- ``authMechanismProperties``: Set to the ``properties`` dictionary that you created in the
  previous step.

You can set these options in two ways: by passing arguments to the
``MongoClient`` constructor or through parameters in your connection string.

> After you set the preceding environment variable, 

Author: good catch. there was so much copy/pasting for this ticket

.. tabs::

   .. tab:: MongoClient
      :tabid: mongoclient

      .. literalinclude:: /includes/authentication/aws-eks-mongoclient.py
            :language: python
            :copyable: true

   .. tab:: Connection String
      :tabid: connectionstring

      .. literalinclude:: /includes/authentication/aws-eks-connection-string.py
            :language: python
            :copyable: true

.. _pymongo-mongodb-oidc-azure-envs:

Other Azure Environments
~~~~~~~~~~~~~~~~~~~~~~~~

If your application runs on Azure Functions, App Service Environment (ASE), or Azure
Kubernetes Service (AKS), you can use the
`azure-identity <https://pypi.org/project/azure-identity/>`__ package to fetch
authentication credentials.

First, use pip to install the ``azure-identity`` library, as shown in the
following example:

.. code-block:: sh

   python3 -m pip install azure-identity

Next, define a class that inherits from the ``OIDCCallback`` class. This class must

> implement one method, ``fetch()``, that returns the OIDC token in the form of an

Reviewer: ```suggestion
implement a ``fetch()`` method, which returns the OIDC token in the form of an
```

   :end-before: client = MongoClient(

Finally, set the following connection options:

- ``authMechanism``: Set to ``"MONGODB-OIDC"``.
- ``authMechanismProperties``: Set to the ``properties`` dictionary that you created in the
  previous step.

You can set these options in two ways: by passing arguments to the
``MongoClient`` constructor or through parameters in your connection string.

.. tabs::

   .. tab:: MongoClient
      :tabid: mongoclient

      .. literalinclude:: /includes/authentication/azure-envs-mongoclient.py
            :language: python
            :copyable: true

   .. tab:: Connection String
      :tabid: connectionstring

      .. literalinclude:: /includes/authentication/azure-envs-connection-string.py
            :language: python
            :copyable: true

.. _pymongo-mongodb-oidc-gcp-gke:

GCP GKE
~~~~~~~

If your application runs on a GCP Google Kubernetes Engine (GKE) cluster with a
`configured service account <https://cloud.google.com/kubernetes-engine/docs/how-to/service-accounts>`__,
you can read the OIDC token from the standard service-account token-file location.

First, define a class that inherits from the ``OIDCCallback`` class. This class must

> implement one method, ``fetch()``, that returns the OIDC token in the form of an

Reviewer: ```suggestion
implement a ``fetch()`` method, which returns the OIDC token in the form of an
```
